File: src/pmskit/rules.py
# -*- coding: utf-8 -*-
"""Modular JSON rule engine (rule-master v1) - phase 1-3 of the rule-engine
migration. Evaluates declarative condition trees (a small, safe JSONLogic-like
dialect - never eval()) over a per-class context with engine-derived fields.

Shadow status: pmskit.validate remains the default engine. run_conventions()
must produce byte-identical findings; tests/test_rules_shadow.py enforces
parity on fixtures. Flip the default only after the parity gate has been green
in CI (see docs/rules.md).

Supported condition operators:
  {"var": "a.b"}                dotted context lookup (None if missing)
  {"truthy": X}                 bool(X)
  {"!": X}                      not X
  {"and": [..]} {"or": [..]}
  {"==":[a,b]} {"!=":[a,b]} {"<":[a,b]} {"<=":[a,b]} {">":[a,b]} {">=":[a,b]}
  {"in":[a,LIST]} {"!in":[a,LIST]}
  {"regex_match":[text, pattern]}   re.search, case-insensitive
  {"facts+": [name, ...]}       union of fact lists (from the pack's "facts")
  {"facts+var": [name, path]}   fact list plus one context value
  {"object": {...}}             literal object with resolved values (contexts)
"""
from __future__ import annotations
import json
import os
import re
from typing import Any, Dict, List, Optional

_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

_CLASS_RE = re.compile(r"\bCL\.?\s*(\d{2,4})\b", re.I)
_FACING_RE = re.compile(r"\b(RTJ|FFS|RJ|RF|FF|MFF|LMF|SMF)\b", re.I)


def load_rulepack(path: Optional[str] = None) -> Dict[str, Any]:
    path = path or os.path.join(_ROOT, "rules", "conventions.json")
    with open(path, encoding="utf-8") as f:
        return json.load(f)


# ---- context building (parsing/derivation lives in code, policy in rules) --
def _to_floats(arr):
    out = []
    for v in arr:
        try:
            out.append(float(v))
        except (TypeError, ValueError):
            return []
    return out


def class_context(c: Dict[str, Any]) -> Dict[str, Any]:
    flange = c.get("flange_rating_face") or ""
    m = _CLASS_RE.search(flange)
    fm = _FACING_RE.search(flange)
    t, p = c.get("temp_C"), c.get("press_barg")
    pt_len_equal = None
    if t and p:
        pt_len_equal = len(t) == len(p)
    mono = False
    if t and p and len(t) == len(p):
        pv = _to_floats(p)
        mono = bool(pv) and any(b - a > 1e-9 for a, b in zip(pv, pv[1:]))
    ctx = dict(c)
    ctx["derived"] = {
        "flange_class_number": int(m.group(1)) if m else None,
        "facing": fm.group(1).upper() if fm else None,
        "has_cl_token": bool(re.search(r"CL", flange, re.I)),
        "pt_len_equal": pt_len_equal,
        "pt_len": len(t) if t else 0,
        "press_len": len(p) if p else 0,
        "press_monotonic_violation": mono,
    }
    return ctx


def component_context(ctx: Dict[str, Any], comp: Dict[str, Any]) -> Dict[str, Any]:
    sub = dict(ctx)
    sub["comp"] = comp
    cm = _CLASS_RE.search(comp.get("description") or "")
    sub["derived"] = dict(ctx["derived"])
    sub["derived"]["component_class_number"] = int(cm.group(1)) if cm else None
    return sub


# ---- expression evaluation --------------------------------------------------
def _var(ctx, path):
    cur = ctx
    for part in path.split("."):
        if not isinstance(cur, dict) or part not in cur:
            return None
        cur = cur[part]
    return cur


def evaluate(expr, ctx, facts):
    if not isinstance(expr, dict):
        return expr
    (op, arg), = expr.items()
    ev = lambda x: evaluate(x, ctx, facts)
    if op == "var":
        return _var(ctx, arg)
    if op == "truthy":
        return bool(ev(arg))
    if op == "!":
        return not bool(ev(arg))
    if op == "and":
        return all(bool(ev(a)) for a in arg)
    if op == "or":
        return any(bool(ev(a)) for a in arg)
    if op in ("==", "!="):
        a, b = ev(arg[0]), ev(arg[1])
        return (a == b) if op == "==" else (a != b)
    if op in ("<", "<=", ">", ">="):
        a, b = ev(arg[0]), ev(arg[1])
        if a is None or b is None:
            return False
        return {"<": a < b, "<=": a <= b, ">": a > b, ">=": a >= b}[op]
    if op in ("in", "!in"):
        a, lst = ev(arg[0]), ev(arg[1])
        res = a in (lst or [])
        return res if op == "in" else not res
    if op == "regex_match":
        text, pat = ev(arg[0]), arg[1]
        return bool(re.search(pat, text or "", re.I))
    if op == "facts+":
        out = []
        for name in arg:
            out += facts.get(name, [])
        return out
    if op == "facts+var":
        return list(facts.get(arg[0], [])) + [_var(ctx, arg[1])]
    if op == "object":
        return {k: ev(v) for k, v in arg.items()}
    raise ValueError(f"unknown rule operator: {op}")


_TPL = re.compile(r"\{([a-zA-Z0-9_.]+)\}")


def render(template, ctx):
    return _TPL.sub(lambda m: str(_var(ctx, m.group(1))), template)


# ---- runner -----------------------------------------------------------------
def run_rules(data: Dict[str, Any], pack: Optional[Dict[str, Any]] = None) -> List[dict]:
    pack = pack or load_rulepack()
    facts = pack.get("facts", {})
    findings: List[dict] = []
    for c in data.get("classes", []):
        ctx = class_context(c)
        for rule in pack.get("rules", []):
            if rule.get("enabled") is False:
                continue
            scopes = [ctx] if rule.get("scope", "class") == "class" else \
                     [component_context(ctx, x) for x in c.get("components", [])]
            for sctx in scopes:
                if evaluate(rule["when"], sctx, facts):
                    context = (evaluate(rule["context"], sctx, facts)
                               if rule.get("context") else None)
                    findings.append({"class": c.get("class"),
                                     "severity": rule["severity"],
                                     "code": rule["id"],
                                     "message": render(rule["message"], sctx),
                                     "context": context})
    return findings


def run_conventions(data: Dict[str, Any]) -> List[dict]:
    """Rule-engine equivalent of pmskit.validate.validate (shadow mode)."""
    return run_rules(data, load_rulepack())

File: web/rules.js
"use strict";
// JS mirror of pmskit/rules.py (rule-master v1). Same operators, same derived
// context, same message templating - byte-identical findings are enforced by
// tools/check_rules_parity.py (Python <-> Node cross-check).
// No eval(): conditions are a tiny structured dialect evaluated recursively.

const RULE_CLASS_RE = /\bCL\.?\s*(\d{2,4})\b/i;
const RULE_FACING_RE = /\b(RTJ|FFS|RJ|RF|FF|MFF|LMF|SMF)\b/i;

function _toFloats(arr){
  const out=[];
  for(const v of (arr||[])){
    const f=parseFloat(v);
    if(!isFinite(f)) return [];
    out.push(f);
  }
  return out;
}

function classContext(c){
  const flange=c.flange_rating_face||"";
  const m=RULE_CLASS_RE.exec(flange), fm=RULE_FACING_RE.exec(flange);
  const t=c.temp_C, p=c.press_barg;
  let ptLenEqual=null;
  if(t&&t.length&&p&&p.length) ptLenEqual=(t.length===p.length);
  let mono=false;
  if(t&&p&&t.length&&p.length&&t.length===p.length){
    const pv=_toFloats(p);
    mono=pv.length>0 && pv.some((v,i)=>i>0 && v-pv[i-1]>1e-9);
  }
  return {...c, derived:{
    flange_class_number: m?parseInt(m[1],10):null,
    facing: fm?fm[1].toUpperCase():null,
    has_cl_token: /CL/i.test(flange),
    pt_len_equal: ptLenEqual,
    pt_len: t?t.length:0,
    press_len: p?p.length:0,
    press_monotonic_violation: mono}};
}

function componentContext(ctx, comp){
  const cm=RULE_CLASS_RE.exec(comp.description||"");
  return {...ctx, comp,
          derived:{...ctx.derived,
                   component_class_number: cm?parseInt(cm[1],10):null}};
}

function _rvar(ctx, path){
  let cur=ctx;
  for(const part of path.split(".")){
    if(cur===null||typeof cur!=="object"||!(part in cur)) return null;
    cur=cur[part];
  }
  return cur===undefined?null:cur;
}

function ruleEval(expr, ctx, facts){
  if(expr===null||typeof expr!=="object"||Array.isArray(expr)) return expr;
  const op=Object.keys(expr)[0], arg=expr[op];
  const ev=x=>ruleEval(x,ctx,facts);
  switch(op){
    case "var": return _rvar(ctx,arg);
    case "truthy": return !!ev(arg);
    case "!": return !ev(arg);
    case "and": return arg.every(a=>!!ev(a));
    case "or": return arg.some(a=>!!ev(a));
    case "==": return ev(arg[0])===ev(arg[1]);
    case "!=": return ev(arg[0])!==ev(arg[1]);
    case "<": case "<=": case ">": case ">=": {
      const a=ev(arg[0]), b=ev(arg[1]);
      if(a===null||b===null) return false;
      return op==="<"?a<b:op==="<="?a<=b:op===">"?a>b:a>=b;
    }
    case "in": case "!in": {
      const a=ev(arg[0]), lst=ev(arg[1])||[];
      const res=lst.includes(a);
      return op==="in"?res:!res;
    }
    case "regex_match": return new RegExp(arg[1],"i").test(ev(arg[0])||"");
    case "facts+": return arg.reduce((o,n)=>o.concat(facts[n]||[]),[]);
    case "facts+var": return (facts[arg[0]]||[]).concat([_rvar(ctx,arg[1])]);
    case "object": {
      const o={};
      for(const k of Object.keys(arg)) o[k]=ev(arg[k]);
      return o;
    }
    default: throw new Error("unknown rule operator: "+op);
  }
}

function ruleRender(tpl, ctx){
  return tpl.replace(/\{([a-zA-Z0-9_.]+)\}/g,(_,p)=>{
    const v=_rvar(ctx,p);
    return v===null?"None":String(v);
  });
}

function runRules(data, pack){
  const facts=pack.facts||{}, findings=[];
  for(const c of (data.classes||[])){
    const ctx=classContext(c);
    for(const rule of (pack.rules||[])){
      if(rule.enabled===false) continue;
      const scopes=(rule.scope||"class")==="class"
        ? [ctx] : (c.components||[]).map(x=>componentContext(ctx,x));
      for(const sctx of scopes){
        if(ruleEval(rule.when,sctx,facts)){
          findings.push({class:c.class??null, severity:rule.severity,
            code:rule.id, message:ruleRender(rule.message,sctx),
            context: rule.context?ruleEval(rule.context,sctx,facts):null});
        }
      }
    }
  }
  return findings;
}

if (typeof module!=="undefined" && module.exports){
  module.exports={runRules, classContext, ruleEval, ruleRender};
}
